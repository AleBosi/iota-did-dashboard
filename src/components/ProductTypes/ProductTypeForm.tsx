import React, { useState } from "react";
import { ProductType } from "../../models/productType";
import { generateDid } from "../../utils/didUtils";
import { issueVC } from "../../utils/vcHelpers";
import { saveItem } from "../../utils/storageHelpers";

interface Props {
  onCreate?: (type: ProductType) => void;
}

const ProductTypeForm: React.FC<Props> = ({ onCreate }) => {
  const [name, setName] = useState("");
  const [description, setDescription] = useState("");

  const handleSubmit = (e: React.FormEvent) => {
    e.preventDefault();
    const newType: ProductType = {
      id: generateDid(), // usa 'id' come nel modello
      name,
      description,
    };
    // Emissione VC di conformità tipologia prodotto
    const issuer = "did:iota:evm:0x..."; // DID azienda (sostituisci con reale)
    const vc = issueVC<ProductType>(
      ["VerifiableCredential", "ProductTypeCredential"],
      issuer,
      newType
    );
    // Aggiunta proprietà extra con cast
    (newType as any).standardCertVCId = vc.id;

    // Salva su localStorage o passa a parent
    saveItem(`ProductType:${newType.id}`, newType);
    saveItem(`VC:${vc.id}`, vc);
    onCreate?.(newType);
    setName("");
    setDescription("");
  };

  return (
    <form onSubmit={handleSubmit}>
      <input
        placeholder="Nome tipologia"
        value={name}
        onChange={e => setName(e.target.value)}
        required
      />
      <input
        placeholder="Descrizione"
        value={description}
        onChange={e => setDescription(e.target.value)}
      />
      <button type="submit">Crea tipologia</button>
    </form>
  );
};

export default ProductTypeForm;
