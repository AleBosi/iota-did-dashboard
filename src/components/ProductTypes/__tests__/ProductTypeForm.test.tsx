import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import ProductTypeForm from "../ProductTypeForm";

describe("ProductTypeForm", () => {
  const mockOnCreate = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test("renderizza campi nome e descrizione e bottone", () => {
    render(<ProductTypeForm onCreate={mockOnCreate} />);
    expect(screen.getByPlaceholderText(/nome tipologia/i)).toBeInTheDocument();
    expect(screen.getByPlaceholderText(/descrizione/i)).toBeInTheDocument();
    expect(screen.getByRole("button", { name: /crea tipologia/i })).toBeInTheDocument();
  });

  test("non submit se nome tipologia Ã¨ vuoto", async () => {
    render(<ProductTypeForm onCreate={mockOnCreate} />);
    fireEvent.click(screen.getByRole("button", { name: /crea tipologia/i }));
    await waitFor(() => {
      expect(mockOnCreate).not.toHaveBeenCalled();
    });
  });

  test("submit chiama onCreate con i dati corretti e resetta i campi", async () => {
    render(<ProductTypeForm onCreate={mockOnCreate} />);
    fireEvent.change(screen.getByPlaceholderText(/nome tipologia/i), { target: { value: "Tipologia Test" } });
    fireEvent.change(screen.getByPlaceholderText(/descrizione/i), { target: { value: "Descrizione test" } });

    fireEvent.click(screen.getByRole("button", { name: /crea tipologia/i }));

    await waitFor(() => {
      expect(mockOnCreate).toHaveBeenCalledTimes(1);
      const arg = mockOnCreate.mock.calls[0][0];
      expect(arg.name).toBe("Tipologia Test");
      expect(arg.description).toBe("Descrizione test");
      expect(arg.id).toBeDefined();
      expect(typeof arg.id).toBe("string");
    });

    expect(screen.getByPlaceholderText(/nome tipologia/i)).toHaveValue("");
    expect(screen.getByPlaceholderText(/descrizione/i)).toHaveValue("");
  });

  // Snapshot
  test("matcha lo snapshot", () => {
    const { asFragment } = render(<ProductTypeForm onCreate={mockOnCreate} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
