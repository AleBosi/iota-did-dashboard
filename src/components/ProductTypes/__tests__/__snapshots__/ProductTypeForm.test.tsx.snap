// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ProductTypeForm matcha lo snapshot 1`] = `
<DocumentFragment>
  <form>
    <input
      placeholder="Nome tipologia"
      required=""
      value=""
    />
    <input
      placeholder="Descrizione"
      value=""
    />
    <button
      type="submit"
    >
      Crea tipologia
    </button>
  </form>
</DocumentFragment>
`;
